packages/svelte/src/internal/client/proxy/readonly.js

```diff
@@ -0,0 +1,65 @@
+import { define_property, get_descriptor } from '../utils.js';
+
+/**
+ * @template {Record<string | symbol, any>} T
+ * @typedef {T & { [READONLY_SYMBOL]: Proxy<T> }} StateObject
+ */
+
+export const READONLY_SYMBOL = Symbol('readonly');
+
+const object_prototype = Object.prototype;
+const array_prototype = Array.prototype;
+const get_prototype_of = Object.getPrototypeOf;
+const is_frozen = Object.isFrozen;
+
+/**
+ * @template {Record<string | symbol, any>} T
+ * @template {StateObject<T>} U
+ * @param {U} value
+ * @returns {Proxy<U> | U}
+ */
+export function readonly(value) {
+	const proxy = value && value[READONLY_SYMBOL];
+	if (proxy) return proxy;
+
+	if (
+		typeof value === 'object' &&
+		value != null &&
+		!is_frozen(value) &&
+		!(READONLY_SYMBOL in value)
+	) {
+		const prototype = get_prototype_of(value);
+
+		// TODO handle Map and Set as well
+		if (prototype === object_prototype || prototype === array_prototype) {
+			const proxy = new Proxy(value, handler);
+			define_property(value, READONLY_SYMBOL, { value: proxy, writable: false });
+
+			return proxy;
+		}
+	}
+
+	return value;
+}
+
+/** @returns {never} */
+const readonly_error = () => {
+	throw new Error(`Props are read-only, unless used with \`bind:\``);
+};
+
+/** @type {ProxyHandler<StateObject<any>>} */
+const handler = {
+	defineProperty: readonly_error,
+	deleteProperty: readonly_error,
+	set: readonly_error,
+
+	get(target, prop, receiver) {
+		const value = Reflect.get(target, prop, receiver);
+
+		if (!(prop in target)) {
+			return readonly(value);
```

Comment:
Is it possible to apply this wrapper on invoking `set/define/delete` instead? People do reads for more often than writes, so this would result in better performance.

Reply:
No — if `foo` is the thing that's read-only, and we haven't wrapped its children, how do we prevent `foo.bar.baz++`?

Reply:
If there's no way to backtrack the path upon set then no there's indeed no way - nevermind then.
This reminds me though: there should probably be a test that checks that this does not apply to classes, i.e. doing `foo.bar.baz++` should be ok if `bar` is a class instance and `baz` a setter on it.